### PR TITLE
Add @@asyncIterator to ReadableStream

### LINKED ISSUE #778: ReadableStream should be an async iterable
(I'm adding this mostly as a tracking bug after I [asked about it in the wrong place](https://github.com/whatwg/fetch/issues/586)) 

NodeJS is [looking into supporting async iterables](https://github.com/tc39/proposal-async-iteration/issues/74) as a way to stream data, and it would be great if fetch (or the readable stream part of fetch) supported the same interface. This would make easier to move code between nodejs and the browser.

### PR DESCRIPTION
**Pull request by MattiasBuelens**
This PR supersedes #954. So far, I've simply rebased @devsnek's changes onto `master` and fixed any conflicts.

WPT PR: web-platform-tests/wpt#15097

Closes #778

***
<a href="https://whatpr.org/streams/980.html" title="Last updated on Feb 6, 2019, 8:31 AM UTC (e452684)">Preview</a> | <a href="https://whatpr.org/streams/980/0ebe4b0...e452684.html" title="Last updated on Feb 6, 2019, 8:31 AM UTC (e452684)">Diff</a>

### CODE REVIEW

**Comment by MattiasBuelens** [comment]
The WPT submodule hasn't yet been updated, so the tests are currently failing on Travis because of the "unexpected" `getIterator` method.

The reference implementation passes all of the new tests locally. Even [the dreaded `Object.prototype` check](https://github.com/web-platform-tests/wpt/pull/13362#pullrequestreview-162720948) works correctly with the new test runner setup! 😄

File: reference-implementation/lib/readable-stream.js [hunk 1]
```diff
@@ -158,8 +158,56 @@ class ReadableStream {
...
+      const result = ReadableStreamReaderGenericCancel(reader, value);
+      ReadableStreamReaderGenericRelease(reader);
+      return result.then(() => ReadableStreamCreateReadResult(value, true, true));
+    }
+    ReadableStreamReaderGenericRelease(reader);
+    return Promise.resolve(ReadableStreamCreateReadResult(value, true, true));
+  }
+}, AsyncIteratorPrototype);
+
+ReadableStream.prototype[Symbol.asyncIterator] = ReadableStream.prototype.getIterator;
```
Comment (MattiasBuelens) [inline reference-implementation/lib/readable-stream.js]: This causes `ReadableStream.prototype[Symbol.asyncIterator]` to be **enumerable**. However, all the other methods are non-enumerable (which is checked by the tests). It also looks like all other objects with an iterator method (`Array`, `Map`, `URLSearchParams`,...) define it as non-enumerable.

@domenic @ricea Shall I make it non-enumerable in the reference implementation, and add a test for it as well?

Comment (devsnek) [reply reference-implementation/lib/readable-stream.js]: I would say yes

**Comment by MattiasBuelens** [comment]
The tests have found a bug in the spec! 🎉 

[This test](https://github.com/web-platform-tests/wpt/blob/ec6765af7c0cbd53bd8c3e9cfab1e54d5fee8e6a/streams/readable-streams/async-iterator.any.js#L216) fails with the current reference implementation. When calling `s.getReader()`, the stream is still locked to the iterator's reader.

The problem is that `it.return()` is not called whenever a `for await..of` loop **completes normally** (i.e. when `it.next()` resolves with `{done: true}`). A quick snippet to demonstrate:
```js
function testIterator() {
  return {
    [Symbol.asyncIterator]() {
      return {
        async next() {
          console.log('next() called');
          return {done: true};
        },
        async return() {
          console.log('return() called');
        }
      };
    }
  };
}

(async () => {
  console.log('before');
  for await (const chunk of testIterator()) {
    console.log({chunk});
  }
  console.log('after');
})();
```
This logs:
```
before
next() called
after
```
Clearly, the `return()` method was not called.

To fix this, we need to change the `next()` method to release the reader just before returning `{done: true}`. For example, we could change [this line in the reference implementation](https://github.com/whatwg/streams/blob/173af93b61a91032c1d270f2215c0683c95d3857/reference-implementation/lib/readable-stream.js#L184) to the following:
```js
    return ReadableStreamDefaultReaderRead(reader)
      .then(({ value, done }) => {
        if (done) {
          ReadableStreamReaderGenericRelease(reader);
        }
        return ReadableStreamCreateReadResult(value, done);
      });
```
Similarly, the spec text should be updated to transform the returned promise with this fulfillment handler.

@domenic @ricea @devsnek Thoughts?

File: reference-implementation/lib/readable-stream.js (hunk 1 above)
Comment (domenic) [reply reference-implementation/lib/readable-stream.js]: Yeah, non-enumerable please!

**Comment by domenic** [comment]
Yep, agreed, we should fix that! This was my second point in https://github.com/whatwg/streams/issues/778#issuecomment-371717821 .

**Comment by MattiasBuelens** [comment]
> Yep, agreed, we should fix that! This was my second point in [#778 (comment)](https://github.com/whatwg/streams/issues/778#issuecomment-371717821) .

All right, I'll fix it.

Any preference on the style of the `.then` handler? The only other example where `ReadableStreamDefaultReaderRead` is used in the spec is [in step 12 of `ReadableStreamTee`](https://whatpr.org/streams/980.html#readable-stream-tee), where the `{done, value}` tuple is explicitly destructured and asserted before it's used. Do we need to do the same thing here? For example:

> 1. Return the result of <a>transforming</a> ! ReadableStreamDefaultReaderRead(_reader_) with a fulfillment handler which takes the argument _result_ and performs the following steps:
>     1. Assert: Type(_result_) is Object.
>     1. Let _value_ be ? Get(_result_, `"value"`).
>     1. Let _done_ be ? Get(_result_, `"done"`).
>     1. Assert: Type(_done_) is Boolean.
>     1. If _done_ is *true*, perform ! ReadableStreamReaderGenericRelease(_reader_).
>     1. Return ! ReadableStreamCreateReadResult(_value_, _done_, *true*).

Can we just check `result.done` directly and pass it along as-is? Or could that cause issues?

> 1. Return the result of <a>transforming</a> ! ReadableStreamDefaultReaderRead(_reader_, *true*) with a fulfillment handler which takes the argument _result_ and performs the following steps:
>     1. If _result.done_ is *true*, perform ! ReadableStreamReaderGenericRelease(_reader_).
>     1. Return _result_.

**Comment by ricea** [comment]
> Any preference on the style of the .then handler? The only other example where ReadableStreamDefaultReaderRead is used internally in the spec is in step 12 of ReadableStreamTee, where the {done, value} tuple is explicitly destructured and asserted before it's used. Do we need to do the same thing here? 

I think we have to do it the long way. It would be weird if `next()` executed `Object.prototype.then` internally.

**Comment by MattiasBuelens** [comment]
> I think we have to do it the long way. It would be weird if `next()` executed `Object.prototype.then` internally.

Good catch, I didn't think about that. I'll do it the long way. 👍

File: index.bs [hunk 2]
```diff
@@ -796,11 +797,11 @@ option. If <code><a for="underlying source">type</a></code> is set to <code>unde
   </code></pre>
 </div>
 
-<h5 id="rs-asynciterator" method for="ReadableStream">getIterator({ <var>preventCancel</var> = false } = {})</h5>
+<h5 id="rs-get-iterator" method for="ReadableStream">getIterator({ <var>preventCancel</var> } = {})</h5>
```
Comment (MattiasBuelens) [inline index.bs]: I didn't like the `= false` inside the method signature, as that also ended up in the table of contents. `pipeTo` and `pipeThrough` don't specify any default values for individual properties either, so this is more consistent. 🙂

File: index.bs [hunk 3]
```diff
@@ -404,6 +406,9 @@ like
                 { preventClose, preventAbort, preventCancel, signal } = {})
     <a href="#rs-pipe-to">pipeTo</a>(dest, { preventClose, preventAbort, preventCancel, signal } = {})
     <a href="#rs-tee">tee</a>()
+
+    <a href="#rs-get-iterator">getIterator</a>({ preventCancel } = {})
```
Comment (MattiasBuelens) [inline index.bs]: What sort of order are we trying to maintain inside these class definitions?

For now, I've put them in the order as they appear in the spec text. If we're aiming for "constructor, properties, methods, symbols", then `getIterator` should go before `getReader`... but should we then also move the spec text for `getIterator` before that of `getReader`? 🤷‍♂️

File: index.bs (hunk 2 above)
Comment (ricea) [reply index.bs]: Agreed.

File: index.bs (hunk 3 above)
Comment (ricea) [reply index.bs]: I think yes, it has to go before getReader, and yes, that means moving the spec text. Sorry :smile:.

Comment (MattiasBuelens) [reply index.bs]: Fair enough. 😛

File: reference-implementation/lib/readable-stream.js [hunk 4]
```diff
@@ -158,8 +158,70 @@ class ReadableStream {
...
+    const reader = AcquireReadableStreamDefaultReader(this);
+    const iterator = Object.create(ReadableStreamAsyncIteratorPrototype);
+    iterator._asyncIteratorReader = reader;
+    iterator._preventCancel = Boolean(preventCancel);
+    return iterator;
+  }
 }
 
+const AsyncIteratorPrototype = Object.getPrototypeOf(Object.getPrototypeOf(async function* () {}).prototype);
+const ReadableStreamAsyncIteratorPrototype = Object.setPrototypeOf({
```
Comment (MattiasBuelens) [inline reference-implementation/lib/readable-stream.js]: New tests, new bugs. This way of setting up the prototype chain causes `next()` and `return()` to become enumerable properties... 🙄 

I guess I'll just resort to `Object.defineProperty` again? Or perhaps `Object.create` with a property descriptor map.

Comment (MattiasBuelens) [reply reference-implementation/lib/readable-stream.js]: Hmmm... What's the cleanest way of creating a function with `.name === 'return'`? The following does not work, because `return` is a keyword:
```
Object.defineProperty(ReadableStreamAsyncIteratorPrototype, 'return', {
  value: function return(value) { ... },
  enumerable: false,
  writable: true,
  configurable: true
});
```

Comment (domenic) [reply reference-implementation/lib/readable-stream.js]: It's simplest to just use object literals and then do `Object.defineProperty(RSAIP, 'return', { enumerable: false })`

Comment (MattiasBuelens) [reply reference-implementation/lib/readable-stream.js]: I am not proud of this contraption:
```js
const AsyncIteratorPrototype = Object.getPrototypeOf(Object.getPrototypeOf(async function* () {}).prototype);
const ReadableStreamAsyncIteratorMethods = class {
  next() { /* ... */ }
  return(value) { /* ... */}
};
const ReadableStreamAsyncIteratorPrototype = Object.create(AsyncIteratorPrototype, {
  next: Object.getOwnPropertyDescriptor(ReadableStreamAsyncIteratorMethods.prototype, 'next'),
  return: Object.getOwnPropertyDescriptor(ReadableStreamAsyncIteratorMethods.prototype, 'return')
});
```
I can't come up with any other way to have a function whose `name` equals `return` though... Please send help! 😆

Comment (domenic) [reply reference-implementation/lib/readable-stream.js]: The following should work:

```js
const RSAIP = Object.setPrototypeOf({ next() { ... }, return() { ... } }, AIP);
Object.defineProperty(RSAIP, 'next', { enumerable: false });
Object.defineProperty(RSAIP, 'return', { enumerable: false });
```

File: index.bs [hunk 5]
```diff
@@ -399,11 +401,14 @@ like
     get <a href="#rs-locked">locked</a>()
 
     <a href="#rs-cancel">cancel</a>(reason)
-    <a href="#rs-get-reader">getReader</a>()
+    <a href="#rs-get-iterator">getIterator</a>({ preventCancel } = {})
+    <a href="#rs-get-reader">getReader</a>({ mode } = {})
```
Comment (MattiasBuelens) [inline index.bs]: Drive-by fix: `getReader()` was missing its `{ mode }` argument in the class definition.

If you prefer I put this in a separate PR, let me know.

**Comment by MattiasBuelens** [comment]
I've pushed a commit to update WPT to the latest commit of the opened WPT PR, to check whether the new tests also pass on Travis. Once the WPT PR lands, I'll amend the commit to point to the merged commit in WPT.

**Comment by MattiasBuelens** [comment]
Aaaand the build fails on linter errors. Probably because I had to update `eslint` to support `async function*() {}`. 🤦‍♂️

I'll see if the linter config can be updated to better match our coding style. If not, I guess I can just `eslint --fix` and commit. 😛

**Comment by MattiasBuelens** [comment]
And now it's `bikeshed` not knowing how to auto-link `[@@asyncIterator]()`... 🤦‍♂️

File: index.bs [hunk 6]
```diff
@@ -792,6 +814,80 @@ option. If <code><a for="underlying source">type</a></code> is set to <code>unde
   </code></pre>
 </div>
 
+<!-- Bikeshed doesn't let us mark this up correctly: https://github.com/tabatkins/bikeshed/issues/1344 -->
+<h5 id="rs-asynciterator" iterator for="ReadableStream">[@@asyncIterator]({ <var>preventCancel</var> } = {})</h5>
```
Comment (MattiasBuelens) [inline index.bs]: Using `iterator` instead of `method` here stops Bikeshed from checking if the content is a valid method name, while still doing the linking. (`attribute` also works.)

A bit of a hack, but good enough for now until Bikeshed lets us do this properly? 😅

Comment (ricea) [reply index.bs]: Seems fine to me. Also, it looks like you did it on purpose, which is a bonus.

File: index.bs [hunk 7]
```diff
@@ -792,6 +814,80 @@ option. If <code><a for="underlying source">type</a></code> is set to <code>unde
...
+  The <code>@@asyncIterator</code> method is an alias of {{ReadableStream/getIterator()}}.
+</p>
+
+The initial value of the <code>@@asyncIterator</code> method is the same function object as the initial value of the
+{{ReadableStream/getIterator()}} method.
+
+<h3 id="rs-asynciterator-prototype" interface
+lt="ReadableStreamAsyncIteratorPrototype">ReadableStreamAsyncIteratorPrototype</h3>
+
+{{ReadableStreamAsyncIteratorPrototype}} is an ordinary object that is used by {{ReadableStream/[@@asyncIterator]()}} to
```
Comment (ricea) [inline index.bs]: I think we should refer to `getIterator()` here and below, instead of `[@@asyncIterator]`, since the latter is defined as an alias of the former.

File: index.bs [hunk 8]
```diff
@@ -792,6 +814,80 @@ option. If <code><a for="underlying source">type</a></code> is set to <code>unde
...
+    <td>\[[preventCancel]]
+    <td class="non-normative">A boolean value indicating if the stream will be <a lt="cancel a readable
+    stream">canceled</a> when the async iterator's {{ReadableStreamAsyncIteratorPrototype/return()}} method is called
+  </tr>
+</table>
+
+<h4 id="rs-asynciterator-prototype-next" method for="ReadableStreamAsyncIteratorPrototype">next()</h4>
+
+<emu-alg>
+  1. If ! IsReadableStreamAsyncIterator(*this*) is *false*, return <a>a promise rejected with</a> *TypeError* exception.
```
Comment (ricea) [inline index.bs]: Need the word "a" before *TypeError*. Also in step 1. of return(), below.

File: reference-implementation/.eslintrc.json [hunk 9]
```diff
@@ -160,7 +160,16 @@
     "id-blacklist": "off",
     "id-length": "off",
     "id-match": "off",
-    "indent": ["error", 2, { "SwitchCase": 1 }],
+    "indent": ["error", 2, {
+      "SwitchCase": 1,
+      "MemberExpression": 2,
+      "FunctionDeclaration": { "parameters": "first" },
```
Comment (ricea) [inline reference-implementation/.eslintrc.json]: The new rules seem reasonable... @domenic, what do you think?

I don't understand why the indentation of `.catch(rethrowAssertionErrorRejection);` expressions has changed, although they always seemed like a weird special case to me anyway.

Comment (MattiasBuelens) [reply reference-implementation/.eslintrc.json]: The `.catch` indentation is controlled by `MemberExpression`. I could set it to `0` to keep the current style, but then it might look weird in other places...

Alternatively, I can change all of those:
```js
promise.then(() => {
  /* ... */
})
.catch(rethrowAssertionErrorRejection);
```
to:
```js
promise.then(() => {
  /* ... */
}).catch(rethrowAssertionErrorRejection);
```
and avoid the problem entirely. 😛

Comment (ricea) [reply reference-implementation/.eslintrc.json]: > Alternatively, I can change all of those:
> 
> ```js
> promise.then(() => {
>   /* ... */
> })
> .catch(rethrowAssertionErrorRejection);
> ```
> 
> to:
> 
> ```js
> promise.then(() => {
>   /* ... */
> }).catch(rethrowAssertionErrorRejection);
> ```

I like this idea, but let's see what @domenic says. These statements predate my involvement in the project.

File: index.bs (hunk 6 above)
Comment (domenic) [reply index.bs]: We shouldn't be adding attributes that do nothing. I'll work on a better version.

File: reference-implementation/.eslintrc.json (hunk 9 above)
Comment (domenic) [reply reference-implementation/.eslintrc.json]: I'm happy to change in that way!

File: index.bs (hunk 6 above)
Comment (domenic) [reply index.bs]: Oh, `iterator` is a valid Bikeshed definition type. Not bad, I'll take it!

**Comment by domenic** [comment]
LGTM! Will let @ricea do the honors for merging tests and then spec. Thanks so much, @MattiasBuelens and @devsnek!

File: reference-implementation/.eslintrc.json (hunk 9 above)
Comment (MattiasBuelens) [reply reference-implementation/.eslintrc.json]: All right, I'll push one more commit to reformat the chained method calls. 🙂

**Comment by zizifn** [comment]
Sorry for create noise here, does anyone know timeline for browers plan to support this?

**Comment by MattiasBuelens** [comment]
> Sorry for create noise here, does anyone know timeline for browers plan to support this?

We've filed implementation bugs with all major browser vendors, see https://github.com/whatwg/streams/issues/778#issuecomment-461341033. From there, it's up to each individual browser vendor to implement this change as they see fit.

You can star the [Chromium issue](https://bugs.chromium.org/p/chromium/issues/detail?id=929585) and/or vote on the [Firefox bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1734244) in order to let them know that you're interested in this feature. Or, if you're willing to put in the effort, you can contribute an implementation yourself. 😉 (Do post a comment on the bug first to inform them that you're working on it though, so they don't end up doing duplicate work.)
